### PR TITLE
PIGs - Proposals for improvement of Gammapy

### PR DESCRIPTION
**Pull request by cdeil**
So far Gammapy development wasn't very structured, and now we want to improve a bit. This is a proposal to have a PIG process to structure major developments and work for Gammapy, by introducing a process and place for people to share major proposals before starting the work.

Concretely @registerrier and I plan to write a proposal how to implement modeling in Gammapy for now, then discuss it a bit, and then to execute the plan in early 2018. This is a good candidate for a PIG, because it requires discussion and evaluation of options, and also has a significant impact on all Gammapy users and developers. So there should be a conscious plan and up-front open and transparent discussion where everyone can contribute their experience and use cases.

This is the outcome of discussions by @adonath @bkhelifi @joleroi @registerrier @robertazanin and me at the Gammapy meeting today at MPIK Heidelberg.

All - Thoughts?

### CODE REVIEW

File: docs/development/pigs/pig-001.rst [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  design in detail and even explaining which alternatives were
+  considered and why the proposed solution was preferred.
+* Have a conscious decision and make sure all interested parties
+  are aware for things that might be controversial and have long-term
+  effects for Gammapy, e.g. when to drop Python 2 support or
+  the Gammapy release cycle. These PIGs can usually be very short.
+
+Most discussions concerning Gammapy will happen by talking to each other directly
+(calls or face-to-face), or online on the mailing list or Github.
+If you're not sure if you should write a GEP, please don't. Instead bring up the
```
Comment (registerrier) [inline docs/development/pigs/pig-001.rst]: you should write a PIG

File: docs/development/pigs/pig-001.rst [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+.. _PEPs: https://www.python.org/dev/peps/pep-0001/
+.. _NEPs: https://docs.scipy.org/doc/numpy-dev/neps/index.html
+.. _APEs: https://github.com/astropy/astropy-APEs
+.. _JEPs: https://github.com/jupyter/enhancement-proposals
+
+We expect that using PIGs will be rare, but we think they can be useful
+e.g. in the following cases:
+
+* Outline design for something that requires significant work,
+  e.g. on topics lik "Modeling in Gammapy" or "High-level user interface in Gammapy".
```
Comment (registerrier) [inline docs/development/pigs/pig-001.rst]: topics like

File: docs/development/pigs/pig-001.rst [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  These APEs can be rather long, more than one page, explaining the
```
Comment (Bultako) [inline docs/development/pigs/pig-001.rst]: These PIGs

**Comment by cdeil** [comment]
@registerrier @Bultako - Thanks for the review!

I have updated the PIG in 3e9db3e7 implementing your suggestions, as well as fixed some more typos and RST formatting issues. I've also re-read the text one more time myself, and changed a few sentences, and added a short section "final remarks".

All - please have a look now, and either leave a comment, or if you agree with this proposal, please "approve" this PR via the Github user interface. If possible, I would prefer if this can get merged soon, so that I have a place to put PIG-2 on modeling. I might also write one or two more to outline proposals for major work in Gammapy for 2018 before the Gammapy meeting in Paris in February.

File: docs/development/pigs/pig-001.rst [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+.. _pig-001:
+
+**********************************
+PIG 1 - PIG purpose and guidelines
+**********************************
+
+* Author: Christoph Deil
+* Created: December 20, 2017
+* Status: draft
```
Comment (adonath) [inline docs/development/pigs/pig-001.rst]: Maybe add a reference to the initial pull request for the PIG, where most of the discussion took place?

File: docs/development/pigs/pig-001.rst [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+main Gammapy repository, and submitted as pull requests.
+
+Most discussions concerning Gammapy will happen by talking to each other directly
+(calls or face-to-face), or online on the mailing list or Github.
+If you're not sure if you should write a PIG, please don't! Instead bring the
+topic up in discussions first with other Gammapy developers, or on the mailing list,
+to get some initial feedback. This will let you figure out if writing a PIG
+will be helpful or not to realise your proposal.
+
+For now, we don't require a formal structure, apart from adding the author,
```
Comment (adonath) [inline docs/development/pigs/pig-001.rst]: Even if we don't **require** a formal structure I think it might actually help the authors to provide a minimal template, even if it only contains the top section.

File: docs/development/pigs/pig-001.rst [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+of the public discussion. But in unusual cases where disagreement remains,
+the coordination committee will talk to the people involved in the discussion
+with the goal to reach consensus or compromise, and then make the final decision.
+
+PIG status
+==========
+
+PIGs can have a status of:
+
+* "draft" - in draft status, either in the writing or dicussion phase
```
Comment (adonath) [inline docs/development/pigs/pig-001.rst]: Are there formal criteria when the PR with the PIG is merged? Is it kept open until the decision is made? Then a "draft" status would not be needed...

File: docs/development/pigs/pig-001.rst [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+* "withdrawn" - withdrawn by the author
+* "accepted" - accepted by the coordination committee
+* "rejected" - rejected by the coordination committee
+
+Final remarks
+=============
+
+This PIG leaves some points open. Specifically we don't say whether
+PIGs are accepted and merged before or after the actual implementation is done,
```
Comment (adonath) [inline docs/development/pigs/pig-001.rst]: This is somehow weird, if the PIG would be accepted after the actual implementation is done...the correct order is that a PIGs gets accepted and then implemented, no? If it's rejected or withdrawn it won't be implemented.

My suggestion would be to already be a *little* bit clearer about the process. E.g. open a pull request with a PIG, the PR is kept open until a decision about the PIG is made. Once the decision is made the PIG is assigned a status and the PR gets merged for the record. Modifications are possible at any time (as long as the PIG is not fully implemented yet) opening a new PR, where the status of the PIG might be revised once this PR gets merged.

File: docs/development/pigs/pig-001.rst (hunk 4 above)
Comment (cdeil) [reply docs/development/pigs/pig-001.rst]: ok, will do.

File: docs/development/pigs/pig-001.rst (hunk 5 above)
Comment (cdeil) [reply docs/development/pigs/pig-001.rst]: I think the structure is different for different PIGs. I can add a sentence pointing to nice examples of 2-3 PEPs or APEs that people can look at for guideline what a high-quality and well-structured proposal can look like.

File: docs/development/pigs/pig-001.rst (hunk 6 above)
Comment (cdeil) [reply docs/development/pigs/pig-001.rst]: I think there will be cases where the PIG PR is merged, if it just proposes a well-thought out plan. And there's other cases where someone puts a very preliminary idea, then there is some discussion, some prototyping, and the PR stays open for months. This has happened for APEs and PEPs and will happen to us, and I think that's OK. So in that phase, the status "draft" is appropriate. The status names are taken from PEPs, see PEP 1. OK?

File: docs/development/pigs/pig-001.rst (hunk 7 above)
Comment (cdeil) [reply docs/development/pigs/pig-001.rst]: > the correct order is that a PIGs gets accepted and then implemented, no? 

Not necessarily, I think it can go either way. E.g. I think it would be awesome if someone tries to use https://www.tensorflow.org/ or https://github.com/HIPS/autograd for extremely efficient modeling and would be willing to contribute that to Gammapy in some form. But I would probably tell them to go implement it first or at least write an advanced prototype, and the PIG would say open until they come back with that. It's the same for many PEPs, usually Guido will only accept them if there's a full implementation already that is complete and stable and ready to be accepted.

So for Gammapy I think we don't want to commit here to anything before we have experience with PIGs. That's why it's so vague.

@adonath - OK? Or if not, could you please make concrete suggestions for changes to the PIG that you would find helpful or clarifying?

File: docs/development/pigs/pig-001.rst [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+PIG 1 - PIG purpose and guidelines
+**********************************
+
+* Author: Christoph Deil
+* Created: December 20, 2017
+* Status: draft
+
+What is a PIG?
+==============
+
```
Comment (joleroi) [inline docs/development/pigs/pig-001.rst]: *This article is about the design document . For other uses, see `Pig (disambiguation)<https://en.wikipedia.org/wiki/Pig_(disambiguation)>`_.*

File: docs/development/pigs/pig-001.rst (hunk 5 above)
Comment (joleroi) [reply docs/development/pigs/pig-001.rst]: > I can add a sentence pointing to nice examples of 2-3 PEPs or APEs that people can look at for guideline what a high-quality and well-structured proposal can look like.

:+1:

File: docs/development/pigs/pig-001.rst [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+in the discussion!
+
+Please note that Github hides inline comments after they have been edited,
+so we suggest that you use inline comments for minor points like spelling mistakes only.
+Put your main feedback as normal comments in the "Conversation" tab,
+so that for someone reading the discussion later they will see your comment directly.
+
+The final decision on any PIG is made by the Gammapy coordination committee.
+We expect that in most cases, the people participating in the PIG review will
+reach a consensus and the coordination committee will simply follow the outcome
```
Comment (joleroi) [inline docs/development/pigs/pig-001.rst]: ~~simply~~

File: docs/development/pigs/pig-001.rst [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+when things change. We also don't specify how the Gammapy coordination committee
+should make they decision if opinions from the people in that group differ.
+This is intentional. We want to keep the process flexible
+and first gain some experience. The goal of PIGs is to help the Gammapy developer
+team to be more efficient, not to have a rigid or bureaucratic process.
+
+.. _PEPs: https://www.python.org/dev/peps/pep-0001/
+.. _NEPs: https://docs.scipy.org/doc/numpy-dev/neps/index.html
+.. _APEs: https://github.com/astropy/astropy-APEs
+.. _JEPs: https://github.com/jupyter/enhancement-proposals
```
Comment (joleroi) [inline docs/development/pigs/pig-001.rst]: The [APE that introduced APEs](https://github.com/astropy/astropy-APEs/blob/master/APE1.rst) - you might call it MOAAPE (mother of all APEs) - has a last section summarizing the discussion, which in this case apparently was very short. But in general I think it's nice to have such a section, especially to avoid that new people attack a PIG with the same arguments that many others have put forward before. For this PIG there won't be much discussion, but I suggest to put in this PIG that other PIGs should have such a paragraph.

File: docs/development/pigs/pig-001.rst (hunk 4 above)
Comment (cdeil) [reply docs/development/pigs/pig-001.rst]: done

File: docs/development/pigs/pig-001.rst (hunk 5 above)
Comment (cdeil) [reply docs/development/pigs/pig-001.rst]: I've added a section here describing structure a bit and pointing to example APEs and PEPs.

File: docs/development/pigs/pig-001.rst (hunk 6 above)
Comment (cdeil) [reply docs/development/pigs/pig-001.rst]: I've added a section here explaining a little more about the status:

> When a PIG is put up for discussion as a pull request, it should have a status of
"draft". Then once the discussion and review is done, the status will change to
one of "withdrawn", "accepted" or "rejected". The reviewers should add a section
"Decision rationale" with a sentence or paragraph summarising the discussion and
decision on this PIG. Then in any case, the PIG should be merged, even if it's
status is "withdrawn" or "rejected".

@adonath - I think your question about decision making and when a PIG should be merged are already addressed in the paragraphs "PIG review" and "Final remarks" that come before / after this "PIG status" section.

File: docs/development/pigs/pig-001.rst (hunk 8 above)
Comment (cdeil) [reply docs/development/pigs/pig-001.rst]: Added. :-)

File: docs/development/pigs/pig-001.rst (hunk 9 above)
Comment (cdeil) [reply docs/development/pigs/pig-001.rst]: Done.

File: docs/development/pigs/pig-001.rst (hunk 10 above)
Comment (cdeil) [reply docs/development/pigs/pig-001.rst]: Agreed. I've added the suggestion to put a "decision rationale" section at the end of the PIG, as they do for APEs.

**Comment by cdeil** [comment]
Updates addressing comments by @adonath and @joleroi in 9856d2bc . IMO this is now in pretty good shape and ready for final review.

@adonath @joleroi @robertazanin @registerrier @bkhelifi or anyone interested - please have another look at this PR and either leave a comment with a suggestion, or if you think it's OK, please "Approve" the PR. If you've never done that, an explanation where to click is here: https://help.github.com/articles/approving-a-pull-request-with-required-reviews/

**Comment by Bultako** [comment]
This is very nice !
In GitHub we have Labels and Milestones to filter and identify PR and Issues. I wonder if we could make use of these functionalities for tracking devels related to a specific PIG. Someone would have to add a label PIG-XXX to PR/Issues related to that PIG-XXX. In the other hand, links to these devels could be added to the Branches and Pull Requests section (like they do in APEs).
My 2cents.

**Comment by cdeil** [comment]
Concerning the comment by @Bultako - Github now also has projects (see https://help.github.com/articles/about-project-boards/) that can be useful to track larger pieces of work that span multiple issues / pull requests.

So far what I've done is usually to mention previous discussions / issues / pull requests in the description when opening a new issue / pull request (see https://help.github.com/articles/autolinked-references-and-urls/#issues-and-pull-requests). I think that would also be enough for larger pieces of work, and only when something spans more than 5 pull requests and months of work I would make a "project" for it to have an overview. The milestone field is something I always set, labels are usually set, but not really used much since Github search with a good word or two usually yields the issues / pull requests I wanted to find.

Bottom line: concerning labels / projects, I'd suggest we just continue like we have, and leave it mainly up to the people doing the work (the implementer and main reviewer) to use labels / project boards as they see fit. So far the problem is more that we don't have active developers, not keeping track of developments via better labels / project boards.

**Comment by robertazanin** [comment]
v

**Comment by cdeil** [comment]
Thanks everyone for the review and feedback on this!
Let's see how things go in the coming months, if this is useful ...
I think it will.

**Comment by cdeil** [comment]
The docs are now updated and the PIG is visible here: http://docs.gammapy.org/dev/development/pigs/pig-001.html

I've changed the status to "accepted" and added a sentence in the "decision rationale" section at the end via a follow-up commit in master: http://docs.gammapy.org/dev/development/pigs/pig-001.html#decision-rationale In the future, I guess it would be a bit nicer to do this directly via a commit here on the PR before merging.
